dm/pkg/checker/conn_checker.go

```diff
@@ -0,0 +1,155 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package checker
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/pingcap/tiflow/dm/dm/config"
+	"github.com/pingcap/tiflow/dm/pkg/conn"
+	tcontext "github.com/pingcap/tiflow/dm/pkg/context"
+	"github.com/pingcap/tiflow/dm/pkg/log"
+)
+
+type connAmountChecker struct {
+	targetDB *conn.BaseDB
+	stCfgs   []*config.SubTaskConfig
+
+	getConfigConn func(stCfgs []*config.SubTaskConfig) int
+	workerName    string
+	unlimitedConn bool
+}
+
+func newConnAmountChecker(targetDB *conn.BaseDB, stCfgs []*config.SubTaskConfig, fn func(stCfgs []*config.SubTaskConfig) int, workerName string) connAmountChecker {
+	return connAmountChecker{
+		targetDB:      targetDB,
+		stCfgs:        stCfgs,
+		getConfigConn: fn,
+		workerName:    workerName,
+	}
+}
+
+func (c *connAmountChecker) check(ctx context.Context, checkerName string) *Result {
+	result := &Result{
+		Name:  checkerName,
+		Desc:  "check if user-specified amount of connection exceeds database's maximum",
+		State: StateFailure,
+	}
+	baseConn, err := c.targetDB.GetBaseConn(ctx)
+	if err != nil {
+		markCheckError(result, err)
+		return result
+	}
+	var rows *sql.Rows
+	rows, err = baseConn.QuerySQL(tcontext.NewContext(ctx, log.L()), "SHOW GLOBAL VARIABLES LIKE 'max_connections'")
+	if err != nil {
+		markCheckError(result, err)
+		return result
+	}
+	defer rows.Close()
+	var (
+		maxConn  int
+		variable string
+	)
+	for rows.Next() {
+		err = rows.Scan(&variable, &maxConn)
+		if err != nil {
+			markCheckError(result, err)
+			return result
+		}
+	}
+	neededConn := c.getConfigConn(c.stCfgs)
+	switch {
+	case maxConn != 0 && neededConn > maxConn:
```

Comment:
we might should also consider currect connections by SHOW PROCESSSLIST.

Reply:
currect -> current?

The `maxConn` is returned by `SHOW GLOBAL VARIABLES LIKE 'max_connections'`, will it be affected by the `SHOW PROCESSSLIST `?

Reply:
Yes current. if there'are already some connections established, we should take them into account

Reply:
20edb4b